chatkit/src/main/java/com/pusher/chatkit/rooms/RoomStateMachine.kt

```diff
@@ -0,0 +1,290 @@
+
+    override fun onCancel() {
+        subscription?.unsubscribe()
+    }
+
+}
+
+private class JoinTask(machine: RoomStateMachine) : InternalRoomTask(machine) {
+    override fun invoke() = Unit
+    override fun onCancel() = Unit
+}
+
+private class AddMessageTask(machine: RoomStateMachine, val room: Room, val messageText: CharSequence) : InternalRoomTask(machine) {
+
+    var pendingPromise: Promise<Any>? = null
+
+    override fun invoke() {
+        pendingPromise = machine.chatManager.currentUser
+            .mapResult { currentUser -> currentUser.name ?: "???" }
+            .mapResult { userName -> Item.Pending(Item.Details(userName, messageText)) as Item }
+            .recover { error -> Item.Failed(Item.Details(messageText, "Could not find name"), error) }
+            .flatMap { item ->
+                machine.update(machine.whenLoaded { copy(items = listOf(item) + items) })
+                machine.chatManager.messageService(room).sendMessage(messageText)
+                    .mapResult { machine.whenLoaded { copy(items = items - item) } }
+                    .recover { error ->
+                        machine.whenLoaded { copy(items = listOf(Item.Failed(item.details, error)) + items - item) }
+                    }
+            }
+            .onReady { update -> machine.update(update) }
+    }
+
+    override fun onCancel() {
+        pendingPromise?.cancel()
+    }
+
+}
+
+private class RetryTask(machine: RoomStateMachine) : InternalRoomTask(machine) {
+    override fun invoke() = Unit
+    override fun onCancel() = Unit
+}
+
+private class LoadMoreTask(machine: RoomStateMachine, val messageCount: Int) : InternalRoomTask(machine) {
+    override fun invoke() = Unit
+    override fun onCancel() = Unit
+}
+
+internal sealed class RoomAction {
+    data class LoadRoom(val roomId: Int) : RoomAction()
+    data class LoadMessages(val room: Room) : RoomAction()
+    data class Join(val room: Room) : RoomAction()
+    data class AddMessage(val room: Room, val message: CharSequence) : RoomAction()
+    data class LoadMore(val messageCount: Int) : RoomAction()
+    object Retry : RoomAction()
+}
+
+private typealias Actor = (RoomAction) -> RoomTask
+
+sealed class RoomState(val type: Type) {
+
+    class Initial internal constructor(private val handle: Actor) : RoomState(INITIAL) {
+        fun loadRoom(roomId: Int): RoomTask = handle(RoomAction.LoadRoom(roomId))
+    }
+
+    data class Idle internal constructor(val roomId: Int) : RoomState(IDLE)
+
+    data class NoMembership internal constructor(
+        private val handle: Actor,
+        val room: Room
+    ) : RoomState(NO_MEMBERSHIP) {
+        fun join(): RoomTask = handle(RoomAction.Join(room))
+    }
+
+    data class RoomLoaded internal constructor(
+        private val handle: Actor,
+        val room: Room
+    ) : RoomState(ROOM_LOADED) {
+        fun loadMessages() = handle(RoomAction.LoadMessages(room))
+    }
+
+    data class Loaded internal constructor(
```

Comment:
What is this called `Loaded` when the associated `RoomState` appears to be `READY`?